vehicle/mkv/software/air_control/air.c

```diff
@@ -1,499 +1,435 @@
+                return;
+            }
+
+            // Set correct scale for pack voltage
+            pack_voltage = pack_voltage * 10;
+
+            /*
+             * This pattern ensures that we only call get_time() once because we
+             * only want to capture the time that PRECHARGE starts
+             */
+            static bool once = true;
+            if (once) {
+                start_time = get_time();
+                once = false;
+            }
+
+            if (get_time() - start_time < 2000) {
+                rc = get_motor_controller_voltage(&motor_controller_voltage);
+                if (rc != 0) {
+                    set_fault(AIR_FAULT_CAN_MC_TIMEOUT);
+                    once = true;
+                    return;
+                }
+
+                // Set correct scale for MC voltage
+                motor_controller_voltage = motor_controller_voltage / 10;
+
+                if (motor_controller_voltage
+                    > (PRECHARGE_THRESHOLD * pack_voltage)) {
+                    gpio_set_pin(AIR_N_LSD); // Close AIR_N
+                    gpio_clear_pin(PRECHARGE_CTL); // Close precharge relay
+                    once = true;
+                    air_control_critical.air_state = TS_ACTIVE;
+                    return;
+                } else {
+                    return;
+                }
+            } else {
+                // 2000 ms (2sec) have elapsed and the motor controller hasn't
+                // reached the appropriate voltage, so precharge failed
+                once = true;
+                set_fault(AIR_FAULT_PRECHARGE_FAIL);
+                return;
+            }
+        } break;
+        case TS_ACTIVE: {
+            // If any of the shutdown nodes open, the SS_TSMS will trigger as
+            // well, so we can just read that one (it is the last node in the
+            // shutdown circuit.
+            if (!air_control_critical.ss_tsms) {
+                air_control_critical.air_state = DISCHARGE;
+            }
+        } break;
+        case DISCHARGE: {
+            gpio_clear_pin(AIR_N_LSD);
+
+            // If either AIR is still closed, it is welded
+            if (air_control_critical.air_p_status
+                && air_control_critical.air_n_status) {
+                air_control_critical.air_state = FAULT;
+                set_fault(AIR_FAULT_BOTH_AIRS_WELD);
+                return;
+            } else if (air_control_critical.air_p_status) {
+                air_control_critical.air_state = FAULT;
+                set_fault(AIR_FAULT_AIR_P_WELD);
+                return;
+            } else if (air_control_critical.air_n_status) {
+                air_control_critical.air_state = FAULT;
+                set_fault(AIR_FAULT_AIR_N_WELD);
+                return;
+            }
+
+            /*
+             * This pattern ensures that we only call get_time() once because we
+             * only want to capture the time that PRECHARGE starts
```

Comment:
Another small thing with this comment, talks about capturing the time PRECHARGE starts when its recording how long the motorcontroller has been discharging for.

Reply:
Another good catch